### ISSUE TITLE
Add ability to upload folder with upload_file

### ISSUE THREAD

**Issue by nateraw** [huggingface/huggingface_hub#497]
**Is your feature request related to a problem? Please describe.**
In the spirit of moving away from forcing users to deal with git, it would be nice to allow folks to upload entire folders using `upload_file`, instead of having them glob/walk through all the files and call `upload_file` each time.

**Describe the solution you'd like**

```python
from huggingface_hub import upload_file

# Uploads contents of `images/` to `nateraw/some-repo`
upload_file('nateraw/some-repo', './images/')
```
**Describe alternatives you've considered**

Bonus points if we have "compression" options as well (zip, tar, gz, etc). If we do that though, we should be mindful of potentially including decompression options in the proposed `download_file` function #460 

**Additional context**
Wanted to upload a few small folders and found myself having to write more code than I wanted to in order to loop over all my local files and call `upload_file`.

**Comment by osanseviero** [comment]
IMO instead of supporting uploading folders with `upload_file`, I would rather have a separate method `upload_folder` or `upload_files` which allows receiving a directory (or file pattern even!) to upload multiple files.

WDYT @LysandreJik?

**Comment by LysandreJik** [comment]
Why not, if it helps! I've tried to parallelize uploads in the past but it has always resulted in 500 errors from the backend.

Feel free to open a PR @osanseviero!

**Comment by osanseviero** [comment]
Yes, the backend fails as discussed in #707. A v0 could push sequentially, which would be a bit of a mess and slow, but still solve the problem. The discussion in https://github.com/huggingface/moon-landing/issues/2260 might unblock this use case though.

**Comment by Pierrci** [comment]
@osanseviero yes, I think it's probably simpler to wait for the revamp of the file upload system

**Comment by nateraw** [comment]
Closed with #888
